core/common/src/main/java/alluxio/master/AlluxioMasterMonitor.java

```diff
@@ -0,0 +1,52 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.master;
+
+import alluxio.exception.status.UnavailableException;
+import alluxio.retry.ExponentialBackoffRetry;
+import alluxio.util.network.NetworkAddressUtils;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.net.InetSocketAddress;
+import java.util.Arrays;
+import java.util.List;
+
+/**
+ * Alluxio master monitor for inquiring RPC service availability.
+ */
+public final class AlluxioMasterMonitor {
+  private static final Logger LOG = LoggerFactory.getLogger(AlluxioMasterMonitor.class);
+  /**
+   * Starts the Alluxio master monitor.
+   *
+   * @param args command line arguments, should be empty
+   */
+  public static void main(String[] args) {
+    InetSocketAddress address = NetworkAddressUtils.getConnectAddress(
+            NetworkAddressUtils.ServiceType.MASTER_RPC);
+    List<InetSocketAddress> addresses = Arrays.asList(address);
+    MasterInquireClient inquireClient = new PollingMasterInquireClient(addresses, () ->
```

Comment:
it can take a while for the master to start, especially in HA mode where it needs to win a Zookeeper election. This could cause spurious failures with too short a timeout. Currently there isn't a great way to tell programmatically whether a secondary master is running properly. For now, I think we can just make sure that one of the masters comes up correctly. We can do this by building a configuration-based client with `FileSystem.Factory.get()`, and then calling a simple method like `fs.exists(new AlluxioURI("/"))`. The client uses defaults designed so that it will retry for longer than it takes the master to start.

To avoid waiting a long time when the master just completely dies, we could also have a second thread watch each host to make sure the `AlluxioMaster` process doesn't go down.

Reply:
I have a few questions:
(1) What do you mean by configuration based client?
(2) Check if the master of a given host is serving RPC is equivalent to check whether it’s AlluxioMaster process is up?

Reply:
1. `FileSystem.Factory.get()` will read alluxio configuration from system properties and `alluxio-site.properties` to create a client which finds the master in the proper way based on what it finds in the configuration. If `alluxio.zookeeper.enabled=true`, the client will query zookeeper to determine the master address, otherwise it will use `alluxio.master.hostname`.
2. They are different - even when multiple masters are up, only one of them (called the primary) will serve RPCs. The rest are called secondaries - they keep track of master state but don't serve any requests. If the primary becomes unavailable, one of the secondaries will be promoted to primary and start serving RPCs

Reply:
It's there is a programmatic way of doing this process check, or I need to ssh to each of the hosts to check if the ```AlluxioMaster``` process still there?

Reply:
It requires `ssh`, but the ssh can be run from Java to make the logic easier to maintain. There is support for invoking CLI from Java in the `ShellUtils.execCommand` method.

Reply:
Done